src/MagnumPlugins/BasisImporter/BasisImporter.cpp

```diff
@@ -0,0 +1,251 @@
+        default:
+            CORRADE_ASSERT_UNREACHABLE();
+    }
+}
+
+constexpr basist::transcoder_texture_format TranscoderTextureFormat[] = {
+    basist::transcoder_texture_format::cTFETC1,
+    basist::transcoder_texture_format::cTFETC2,
+    basist::transcoder_texture_format::cTFBC1,
+    basist::transcoder_texture_format::cTFBC3,
+    basist::transcoder_texture_format::cTFBC4,
+    basist::transcoder_texture_format::cTFBC5,
+    basist::transcoder_texture_format::cTFBC7_M6_OPAQUE_ONLY,
+    basist::transcoder_texture_format::cTFPVRTC1_4_OPAQUE_ONLY
+};
+
+}
+
+struct BasisImporter::State {
+    /* There is only this type of codebook */
+    basist::etc1_global_selector_codebook _codebook;
+    basist::basisu_transcoder _transcoder{&_codebook};
+
+    State(): _codebook(basist::g_global_selector_cb_size,
+        basist::g_global_selector_cb) {}
+};
+
+void BasisImporter::initialize() {
+    basist::basisu_transcoder_init();
+}
+
+BasisImporter::BasisImporter() = default;
+
+BasisImporter::BasisImporter(PluginManager::AbstractManager& manager, const std::string& plugin): AbstractImporter{manager, plugin} {
+    /* Initializes codebook */
+    _state.reset(new State);
+
+    /* Set format configuration from plugin alias */
+    if(Corrade::Utility::String::beginsWith(plugin, "BasisImporter")) {
+
+        /* Has type prefix */
+        if(plugin.length() > 13) {
+            /* We can assume the substring results in a valid value
+               as the plugin conf limits it to known suffixes */
+            configuration().setValue("format", plugin.substr(13));
+        }
+    }
+}
+
+BasisImporter::~BasisImporter() = default;
+
+auto BasisImporter::doFeatures() const -> Features { return Feature::OpenData; }
+
+bool BasisImporter::doIsOpened() const { return _in; }
+
+void BasisImporter::doClose() { _in = nullptr; }
+
+void BasisImporter::doOpenData(const Containers::ArrayView<const char> data) {
+    /* Because here we're copying the data and using the _in to check if file
+       is opened, having them nullptr would mean openData() would fail without
+       any error message. It's not possible to do this check on the importer
+       side, because empty file is valid in some formats (OBJ or glTF). We also
+       can't do the full import here because then doImage2D() would need to
+       copy the imported data instead anyway (and the uncompressed size is much
+       larger). This way it'll also work nicely with a future openMemory(). */
+    if(data.empty()) {
+        Error{} << "Trade::BasisImporter::openData(): the file is empty";
+        return;
+    }
+
+    if(!_state->_transcoder.validate_header(data.data(), data.size())) {
+        Error() << "Trade::BasisImporter::openData(): invalid header";
+        return;
+    }
+
+    if(!_state->_transcoder.start_transcoding(data.data(), data.size())) {
+        Error() << "Trade::BasisImporter::openData(): bad basis file";
+        return;
+    }
```

Comment:
I think you could craft a `fileTooShort()` test for this one by passing just a prefix of the data to `openData()`.

Reply:
This will fail validate header first independent of how long i leave the prefix. But I verifiy the `invalid header` error now.

Reply:
Ah because there's the CRC check. In what cases does `start_transcoding()`` actually return false?

Reply:
Well, the header check [checks the file size ](https://github.com/BinomialLLC/basis_universal/blob/master/transcoder/basisu_transcoder.cpp#L4226)

I assume one could pass random data after the signature 🤔